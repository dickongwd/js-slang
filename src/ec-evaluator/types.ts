import * as es from 'estree'

import { Context } from '..'
import { Environment, Value } from '../types'
import { Agenda, Stash } from './interpreter'

export type AgendaItem = es.Node | IInstr

// Might have to change each InstrType to its own interface because different instrtypes have different
// additional properties that they require.
export interface IInstr {
  instrType: InstrTypes
  symbol?: string // for Assignment
  constant?: boolean // for Assignment
  declaration?: boolean // for Assignment
  numOfArgs?: number // for Application
  expr?: es.CallExpression // for Application error handling
  env?: Environment // For restoring environments
}

export enum InstrTypes {
  RESET = 'Reset',
  WHILE = 'While',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED_IF_NEEDED = 'PushUndefinedIfNeeded',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  MARKER = 'Marker'
}

export type cmdEvaluator = (
  command: AgendaItem,
  context: Context,
  agenda: Agenda,
  stash: Stash
) => Value

// Tags taken from MicroPython interpreter
// export enum Tags {
//   /**
//    * Source ยง3 expressions
//    */
//   Lit = 'lit',
//   Nam = 'nam',
//   UnOp = 'unop',
//   BinOp = 'binop',
//   Log = 'log',
//   CondExpr = 'cond_expr',
//   App = 'app',
//   Assmt = 'assmt',
//   Lam = 'lam',
//   Spread = 'spread',
//   ArrLit = 'arr_lit',
//   ArrAcc = 'arr_acc',
//   ArrAssmt = 'arr_asmt',

//   /**
//    * Source ยง3 Statements
//    */
//   Import = 'import',
//   Seq = 'seq',
//   CondStmt = 'cond_stmt',
//   Blk = 'blk',
//   Let = 'let',
//   Const = 'const',
//   Ret = 'ret',
//   Fun = 'fun',
//   While = 'while',
//   For = 'for',
//   Prop = 'prop',

//   /**
//    * CSE machine instructions
//    */
//   ResetInstr = 'reset_i',
//   WhileInstr = 'while_i',
//   AssmtInstr = 'assmt_i',
//   UnOpInstr = 'unop_i',
//   BinOpInstr = 'binop_i',
//   PopInstr = 'pop_i',
//   AppInstr = 'app_i',
//   BranchInstr = 'branch_i',
//   EnvInstr = 'env_i',
//   PushUndefInstr = 'push_undefined_if_needed_i',
//   ArrLitInstr = 'arr_lit_i',
//   ArrAccInstr = 'arr_acc_i',
//   ArrAssmtInstr = 'arr_assmt_i'
// }
