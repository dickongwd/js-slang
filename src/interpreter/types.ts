import * as es from 'estree'
import { Context } from '..'
import { Value } from '../types'
import { Agenda, Stash } from './interpreter'

export type AgendaItem = es.Node | IInstr

// Might have to change each InstrType to its own interface because different instrtypes have different
// additional properties they need to store.
export interface IInstr {
  instrType: InstrTypes
  symbol? : string
  const? : boolean
}

export enum InstrTypes {
  RESET = 'Reset',
  WHILE = 'While',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  PUSH_UNDEFINED = 'PushUndefined',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ASSIGNMENT = 'ArrayAssignment'
}

export type cmdEvaluator = (command: AgendaItem, context: Context, agenda: Agenda, stash: Stash) => Value

// Tags taken from MicroPython interpreter
// export enum Tags {
//   /**
//    * Source ยง3 expressions
//    */
//   Lit = 'lit',
//   Nam = 'nam',
//   UnOp = 'unop',
//   BinOp = 'binop',
//   Log = 'log',
//   CondExpr = 'cond_expr',
//   App = 'app',
//   Assmt = 'assmt',
//   Lam = 'lam',
//   Spread = 'spread',
//   ArrLit = 'arr_lit',
//   ArrAcc = 'arr_acc',
//   ArrAssmt = 'arr_asmt',

//   /**
//    * Source ยง3 Statements
//    */
//   Import = 'import',
//   Seq = 'seq',
//   CondStmt = 'cond_stmt',
//   Blk = 'blk',
//   Let = 'let',
//   Const = 'const',
//   Ret = 'ret',
//   Fun = 'fun',
//   While = 'while',
//   For = 'for',
//   Prop = 'prop',

//   /**
//    * CSE machine instructions
//    */
//   ResetInstr = 'reset_i',
//   WhileInstr = 'while_i',
//   AssmtInstr = 'assmt_i',
//   UnOpInstr = 'unop_i',
//   BinOpInstr = 'binop_i',
//   PopInstr = 'pop_i',
//   AppInstr = 'app_i',
//   BranchInstr = 'branch_i',
//   EnvInstr = 'env_i',
//   PushUndefInstr = 'push_undefined_if_needed_i',
//   ArrLitInstr = 'arr_lit_i',
//   ArrAccInstr = 'arr_acc_i',
//   ArrAssmtInstr = 'arr_assmt_i'
// }
