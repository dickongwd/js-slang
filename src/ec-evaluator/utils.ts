import * as es from 'estree'

import {} from '../utils/astCreator'
import { popInstr } from './instrCreator'
import { AgendaItem } from './types'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(...items: T[]): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  private storage: T[] = []

  public constructor() {}

  public push(...items: T[]): void {
    for (const item of items) {
      this.storage.push(item)
    }
  }

  public pop(): T | undefined {
    return this.storage.pop()
  }

  public peek(): T | undefined {
    if (this.size() === 0) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  public size(): number {
    return this.storage.length
  }
}

/**
 * Typeguard for esNode to distinguish between program statements and instructions.
 *
 * @param command An AgendaItem
 * @returns true if the AgendaItem is an esNode and false if it is an instruction.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}

/**
 * A helper function for handling sequences of statements.
 * Statements must be pushed in reverse order, and each statement is separated by a pop
 * instruction so that only the result of the last statement remains on stash.
 * Value producing statements have an extra pop instruction.
 *
 * @param seq Array of statements.
 * @returns Array of commands to be pushed into agenda.
 */
export const handleSequence = (seq: es.Statement[]): AgendaItem[] => {
  const result: AgendaItem[] = []
  let valueProduced = false
  for (const command of seq) {
    if (valueProducing(command)) {
      valueProduced ? result.push(popInstr()) : (valueProduced = true)
    }
    result.push(command)
  }
  return result.reverse()
}

/**
 * To determine if an agenda item is value producing. JavaScript distinguishes value producing
 * statements and non-value producing statements.
 * Refer to https://sourceacademy.nus.edu.sg/sicpjs/4.1.2 exercise 4.8.
 *
 * @param command Agenda item to determine if it is value producing.
 * @returns true if it is value producing, false otherwise.
 */
export const valueProducing = (command: es.Node): boolean => {
  const type = command.type
  return (
    type !== 'VariableDeclaration' &&
    type !== 'FunctionDeclaration' &&
    (type !== 'BlockStatement' || command.body.some(valueProducing))
  )
}
