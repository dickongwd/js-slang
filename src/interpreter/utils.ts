import { AgendaItem} from './types'
import * as es from 'estree'

/**
 * Stack is implemented for agenda and stash registers.
 */
interface IStack<T> {
  push(item: T): void
  pop(): T | undefined
  peek(): T | undefined
  size(): number
}

export class Stack<T> implements IStack<T> {
  // Bottom of the array is at index 0
  protected storage: T[] = []

  constructor() {}

  push(item: T) {
    this.storage.push(item)
  }

  extend(items: T[]) {
    this.storage.push(...items)
  }

  pop() {
    return this.storage.pop()
  }

  peek() {
    if (this.size() === 0) {
      return undefined
    }
    return this.storage[this.size() - 1]
  }

  size() {
    return this.storage.length
  }
}

/**
 * Typeguard for esNode
 * @param command an AgendaItem
 * @returns true if the AgendaItem is an esNode and false otherwise.
 */
export const isNode = (command: AgendaItem): command is es.Node => {
  return (command as es.Node).type !== undefined
}
