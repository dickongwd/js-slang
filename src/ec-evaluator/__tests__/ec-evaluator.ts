import { Chapter, Variant } from '../../types'
import { stripIndent } from '../../utils/formatters'
import { expectParsedError, expectParsedErrorNoSnapshot, expectResult } from '../../utils/testing'

test('Check that stack is at most 10k in size', () => {
  return expectParsedErrorNoSnapshot(
    stripIndent`
    function f(x) {
      if (x <= 0) {
        return 0;
      } else {
        return 1 + f(x-1);
      }
    }
    f(10000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toEqual(expect.stringMatching(/Maximum call stack size exceeded\n([^f]*f){3}/))
}, 10000)

test('Simple tail call returns work', () => {
  return expectResult(
    stripIndent`
    function f(x, y) {
      if (x <= 0) {
        return y;
      } else {
        return f(x-1, y+1);
      }
    }
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail call in conditional expressions work', () => {
  return expectResult(
    stripIndent`
    function f(x, y) {
      return x <= 0 ? y : f(x-1, y+1);
    }
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail call in boolean operators work', () => {
  return expectResult(
    stripIndent`
    function f(x, y) {
      if (x <= 0) {
        return y;
      } else {
        return false || f(x-1, y+1);
      }
    }
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail call in nested mix of conditional expressions boolean operators work', () => {
  return expectResult(
    stripIndent`
    function f(x, y) {
      return x <= 0 ? y : false || x > 0 ? f(x-1, y+1) : 'unreachable';
    }
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail calls in arrow functions work', () => {
  return expectResult(
    stripIndent`
    const f = (x, y) => x <= 0 ? y : f(x-1, y+1);
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail calls in arrow block functions work', () => {
  return expectResult(
    stripIndent`
    const f = (x, y) => {
      if (x <= 0) {
        return y;
      } else {
        return f(x-1, y+1);
      }
    };
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail calls in mutual recursion work', () => {
  return expectResult(
    stripIndent`
    function f(x, y) {
      if (x <= 0) {
        return y;
      } else {
        return g(x-1, y+1);
      }
    }
    function g(x, y) {
      if (x <= 0) {
        return y;
      } else {
        return f(x-1, y+1);
      }
    }
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail calls in mutual recursion with arrow functions work', () => {
  return expectResult(
    stripIndent`
    const f = (x, y) => x <= 0 ? y : g(x-1, y+1);
    const g = (x, y) => x <= 0 ? y : f(x-1, y+1);
    f(5000, 5000);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`10000`)
})

test('Tail calls in mixed tail-call/non-tail-call recursion work', () => {
  return expectResult(
    stripIndent`
    function f(x, y, z) {
      if (x <= 0) {
        return y;
      } else {
        return f(x-1, y+f(0, z, 0), z);
      }
    }
    f(5000, 5000, 2);
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`15000`)
})

// This is bad practice. Don't do this!
test('standalone block statements', () => {
  return expectResult(
    stripIndent`
    function test(){
      const x = true;
      {
          const x = false;
      }
      return x;
    }
    test();
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`true`)
})

// This is bad practice. Don't do this!
test('const uses block scoping instead of function scoping', () => {
  return expectResult(
    stripIndent`
    function test(){
      const x = true;
      if(true) {
          const x = false;
      } else {
          const x = false;
      }
      return x;
    }
    test();
  `,
    { variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`true`)
})

// This is bad practice. Don't do this!
test('let uses block scoping instead of function scoping', () => {
  return expectResult(
    stripIndent`
    function test(){
      let x = true;
      if(true) {
          let x = false;
      } else {
          let x = false;
      }
      return x;
    }
    test();
  `,
    { chapter: Chapter.SOURCE_3, variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`true`)
})

// This is bad practice. Don't do this!
xtest('for loops use block scoping instead of function scoping', () => {
  return expectResult(
    stripIndent`
    function test(){
      let x = true;
      for (let x = 1; x > 0; x = x - 1) {
      }
      return x;
    }
    test();
  `,
    { chapter: Chapter.SOURCE_3, variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`true`)
})

xtest('while loops use block scoping instead of function scoping', () => {
  return expectResult(
    stripIndent`
    function test(){
      let x = true;
      while (true) {
        let x = false;
        break;
      }
      return x;
    }
    test();
  `,
    { chapter: Chapter.SOURCE_4, variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`true`)
})

// see https://www.ecma-international.org/ecma-262/6.0/#sec-for-statement-runtime-semantics-labelledevaluation
// and https://hacks.mozilla.org/2015/07/es6-in-depth-let-and-const/
xtest('for loop `let` variables are copied into the block scope', () => {
  return expectResult(
    stripIndent`
  function test(){
    let z = [];
    for (let x = 0; x < 10; x = x + 1) {
      z[x] = () => x;
    }
    return z[1]();
  }
  test();
  `,
    { chapter: Chapter.SOURCE_4, variant: Variant.EXPLICIT_CONTROL }
  ).toMatchInlineSnapshot(`1`)
})

test('Cannot overwrite loop variables within a block', () => {
  return expectParsedError(
    stripIndent`
  function test(){
      let z = [];
      for (let x = 0; x < 2; x = x + 1) {
        x = 1;
      }
      return false;
  }
  test();
  `,
    { chapter: Chapter.SOURCE_3 }
  ).toMatchInlineSnapshot(
    `"Line 4: Assignment to a for loop variable in the for loop is not allowed."`
  )
})

test('No hoisting of functions. Only the name is hoisted like let and const', () => {
  return expectParsedError(stripIndent`
      const v = f();
      function f() {
        return 1;
      }
      v;
    `).toMatchInlineSnapshot(
    `"Line 1: Name f declared later in current scope but not yet assigned"`
  )
}, 30000)

test('Error when accessing temporal dead zone', () => {
  return expectParsedError(stripIndent`
    const a = 1;
    function f() {
      display(a);
      const a = 5;
    }
    f();
    `).toMatchInlineSnapshot(
    `"Line 3: Name a declared later in current scope but not yet assigned"`
  )
}, 30000)

// tslint:disable-next-line:max-line-length
test('In a block, every going-to-be-defined variable in the block cannot be accessed until it has been defined in the block.', () => {
  return expectParsedError(stripIndent`
      const a = 1;
      {
        a + a;
        const a = 10;
      }
    `).toMatchInlineSnapshot(
    `"Line 3: Name a declared later in current scope but not yet assigned"`
  )
}, 30000)

test('Shadowed variables may not be assigned to until declared in the current scope', () => {
  return expectParsedError(
    stripIndent`
  let variable = 1;
  function test(){
    variable = 100;
    let variable = true;
    return variable;
  }
  test();
  `,
    { chapter: Chapter.SOURCE_3 }
  ).toMatchInlineSnapshot(`"Line 3: Name variable not declared."`)
})
