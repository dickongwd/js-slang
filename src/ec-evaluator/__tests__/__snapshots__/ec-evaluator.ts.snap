// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cannot overwrite loop variables within a block: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "function test(){
    let z = [];
    for (let x = 0; x < 2; x = x + 1) {
      x = 1;
    }
    return false;
}
test();",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 4: Assignment to a for loop variable in the for loop is not allowed.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Error when accessing temporal dead zone: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = 1;
function f() {
  display(a);
  const a = 5;
}
f();",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 3: Name a declared later in current scope but not yet assigned",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`In a block, every going-to-be-defined variable in the block cannot be accessed until it has been defined in the block.: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "const a = 1;
{
  a + a;
  const a = 10;
}",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 3: Name a declared later in current scope but not yet assigned",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`No hoisting of functions. Only the name is hoisted like let and const: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "const v = f();
function f() {
  return 1;
}
v;",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 1: Name f declared later in current scope but not yet assigned",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Shadowed variables may not be assigned to until declared in the current scope: expectParsedError 1`] = `
Object {
  "alertResult": Array [],
  "code": "let variable = 1;
function test(){
  variable = 100;
  let variable = true;
  return variable;
}
test();",
  "displayResult": Array [],
  "numErrors": 1,
  "parsedErrors": "Line 3: Name variable not declared.",
  "result": undefined,
  "resultStatus": "error",
  "visualiseListResult": Array [],
}
`;

exports[`Simple tail call returns work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y) {
  if (x <= 0) {
    return y;
  } else {
    return f(x-1, y+1);
  }
}
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail call in boolean operators work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y) {
  if (x <= 0) {
    return y;
  } else {
    return false || f(x-1, y+1);
  }
}
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail call in conditional expressions work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y) {
  return x <= 0 ? y : f(x-1, y+1);
}
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail call in nested mix of conditional expressions boolean operators work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y) {
  return x <= 0 ? y : false || x > 0 ? f(x-1, y+1) : 'unreachable';
}
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail calls in arrow block functions work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const f = (x, y) => {
  if (x <= 0) {
    return y;
  } else {
    return f(x-1, y+1);
  }
};
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail calls in arrow functions work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const f = (x, y) => x <= 0 ? y : f(x-1, y+1);
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail calls in mixed tail-call/non-tail-call recursion work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y, z) {
  if (x <= 0) {
    return y;
  } else {
    return f(x-1, y+f(0, z, 0), z);
  }
}
f(5000, 5000, 2);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 15000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail calls in mutual recursion with arrow functions work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const f = (x, y) => x <= 0 ? y : g(x-1, y+1);
const g = (x, y) => x <= 0 ? y : f(x-1, y+1);
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Tail calls in mutual recursion work: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function f(x, y) {
  if (x <= 0) {
    return y;
  } else {
    return g(x-1, y+1);
  }
}
function g(x, y) {
  if (x <= 0) {
    return y;
  } else {
    return f(x-1, y+1);
  }
}
f(5000, 5000);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 10000,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`const uses block scoping instead of function scoping: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function test(){
  const x = true;
  if(true) {
      const x = false;
  } else {
      const x = false;
  }
  return x;
}
test();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`let uses block scoping instead of function scoping: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function test(){
  let x = true;
  if(true) {
      let x = false;
  } else {
      let x = false;
  }
  return x;
}
test();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`standalone block statements: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "function test(){
  const x = true;
  {
      const x = false;
  }
  return x;
}
test();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
